File: src/main/java/me/lokka30/levelledmobs/rules/TieredColoringInfo.java
/*
 * Copyright (c) 2020-2021  lokka30. Use of this source code is governed by the GNU AGPL v3.0 license that can be found in the LICENSE.md file.
 */

package me.lokka30.levelledmobs.rules;

import me.lokka30.levelledmobs.util.Utils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Holds the values parsed from rules.yml used with the tiered placeholder for nametags
 *
 * @author stumper66
 * @since 3.0.0
 */
public class TieredColoringInfo implements Cloneable {

    public int minLevel;
    public int maxLevel;
    public String text;
    boolean isDefault;

    private TieredColoringInfo() {
    }

    @NotNull
    static TieredColoringInfo createDefault(final @NotNull String value) {
        final TieredColoringInfo coloringInfo = new TieredColoringInfo();
        coloringInfo.isDefault = true;
        coloringInfo.text = value;

        return coloringInfo;
    }

    @Nullable
    static TieredColoringInfo createFromString(final @NotNull String key,
        final @NotNull String value) {
        final String[] numbers = key.split("-");
        if (numbers.length != 2) {
            Utils.logger.warning("Invalid tiered coloring key: " + key);
            return null;
        }

        final TieredColoringInfo coloringInfo = new TieredColoringInfo();

        for (int i = 0; i < 2; i++) {
            final String num = numbers[i].trim();
            if (!Utils.isInteger(num)) {
                Utils.logger.warning("Invalid number in tiered coloring key: " + key);
                return null;
            }

            if (i == 0) {
                coloringInfo.minLevel = Integer.parseInt(num);
            } else {
                coloringInfo.maxLevel = Integer.parseInt(num);
            }
        }

        coloringInfo.text = value;

        return coloringInfo;
    }

    public String toString() {
        if (isDefault) {
            return this.text + "default&r";
        } else {
            return String.format("%s%s-%s&r", this.text, this.minLevel, this.maxLevel);
        }
    }

    public TieredColoringInfo cloneItem() {
        TieredColoringInfo copy = null;
        try {
            copy = (TieredColoringInfo) super.clone();
        } catch (final Exception e) {
            e.printStackTrace();
        }

        return copy;
    }
}

File: levelledmobs-plugin/src/main/kotlin/io/github/arcaneplugins/levelledmobs/rules/TieredColoringInfo.kt
package io.github.arcaneplugins.levelledmobs.rules

import io.github.arcaneplugins.levelledmobs.util.Utils
import io.github.arcaneplugins.levelledmobs.util.Utils.isInteger

/**
 * Holds the values parsed from rules.yml used with the tiered placeholder for nametags
 *
 * @author stumper66
 * @since 3.0.0
 */
class TieredColoringInfo : Cloneable {
    var minLevel = 0
    var maxLevel = 0
    var text: String? = null
    var isDefault = false

    companion object{
        fun createDefault(value: String): TieredColoringInfo {
            val coloringInfo = TieredColoringInfo()
            coloringInfo.isDefault = true
            coloringInfo.text = value

            return coloringInfo
        }

        fun createFromString(
            key: String,
            value: String
        ): TieredColoringInfo? {
            val numbers = key.split("-".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

            if (numbers.size != 2 && numbers.size != 1) {
                Utils.logger.warning("Invalid tiered coloring key: $key")
                return null
            }

            val coloringInfo = TieredColoringInfo()
            coloringInfo.text = value

            if (numbers.size == 1) {
                coloringInfo.minLevel = numbers[0].toInt()
                coloringInfo.maxLevel = coloringInfo.minLevel
                return coloringInfo
            }

            for (i in 0..1) {
                val num = numbers[i].trim { it <= ' ' }
                if (!isInteger(num)) {
                    Utils.logger.warning("Invalid number in tiered coloring key: $key")
                    return null
                }

                if (i == 0) {
                    coloringInfo.minLevel = num.toInt()
                } else {
                    coloringInfo.maxLevel = num.toInt()
                }
            }

            return coloringInfo
        }
    }

    override fun toString(): String {
        return if (isDefault) {
            text + "default&r"
        } else {
            "$text$minLevel-$maxLevel&r"
        }
    }

    fun cloneItem(): TieredColoringInfo? {
        var copy: TieredColoringInfo? = null
        try {
            copy = super.clone() as TieredColoringInfo?
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return copy
    }
}